File: hy-poc-flink/src/main/scala/com/hy/poc/flink/datastream/sink/StreamingFileSinkPoc.scala
package com.hy.poc.flink.datastream.sink

import org.apache.flink.api.common.serialization.SimpleStringEncoder
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.core.fs.Path
import org.apache.flink.runtime.state.hashmap.HashMapStateBackend
import org.apache.flink.streaming.api.environment.{CheckpointConfig, StreamExecutionEnvironment}
import org.apache.flink.streaming.api.functions.sink.filesystem.{OutputFileConfig, StreamingFileSink}
import org.apache.flink.streaming.api.functions.sink.filesystem.bucketassigners.DateTimeBucketAssigner
import org.apache.flink.streaming.api.functions.sink.filesystem.rollingpolicies.DefaultRollingPolicy
import org.slf4j.{Logger, LoggerFactory}
import org.apache.flink.streaming.api.datastream.DataStreamSource

/**
  * SocketHelloWorld
  *
  * @author Jie.Hu
  * @date 7/12/21 12:57 PM
  */
object StreamingFileSinkPoc {

  private val logger: Logger = LoggerFactory.getLogger(StreamingFileSinkPoc.getClass)

  def main(args: Array[String]): Unit = {

    val tool = ParameterTool.fromArgs(args)
    val streamEnv = StreamExecutionEnvironment.getExecutionEnvironment
    val socketStream: DataStreamSource[String] = streamEnv.socketTextStream(tool.get("host"), tool.getInt("port"))

    val config = streamEnv.getConfig
    val checkpointCfg = streamEnv.getCheckpointConfig
    checkpointCfg.enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION)
    streamEnv.enableCheckpointing(tool.getLong("checkpoint.interval"))
    streamEnv.setStateBackend(new HashMapStateBackend())
    checkpointCfg.setCheckpointStorage(tool.get("checkpoint.dir"))

    val fileSink = StreamingFileSink
      .forRowFormat(new Path(tool.get("output")), new SimpleStringEncoder[String]())
      // bucket 分配器，即如何分配 bucket，默认根据小时分 bucket
      .withBucketAssigner(new DateTimeBucketAssigner())
      .withBucketCheckInterval(1000L)
      .withOutputFileConfig(
        // 输出文件配置
        OutputFileConfig.builder()
          .withPartPrefix("poc_row")
          .withPartSuffix(".csv")
          .build())
      .withRollingPolicy(
        // note that:
        //    只有打开了 checkpoint 才会进行回滚
        //    如果 checkpoint 被禁用了，part 文件将永远处于 in-progress 或 pending 状态，下游无法安全使用数据
        // 默认的回滚策略如下
        //    1. 如果 当前时间 - 文件的创建时间 > RolloverInterval 就关闭 inprogress 文件
        //    2. 如果 当前的时间 - 文件的最后修改时间 > InactivityInterval 就关闭 inprogress 文件
        //    3. 如果 文件的大小 > max part size 就关闭 inprogress 文件
        // 关闭 inprogress 文件后，就将文件改成 pending 状态，等到下一次 checkpoint 成功之后，将文件改为 finished 状态
        // 请注意，修改状态需要开启 checkpoint ，否则不会修改文件状态，因为只有在 checkpoint 时才会调用方法
        // 正确的文件回滚策略应该是
        //    1. 确认多久生成一个文件，即 RolloverInterval，这个时间应该小于等于 checkpoint
        //    2. 确认如果没有新数据写入的时候应该多久关闭文件，即InactivityInterval，这个时间应该小于 RolloverInterval
        //    3. 确认文件的大小，默认为 128M
        DefaultRollingPolicy.builder()
          // 文件不活动时间
          .withInactivityInterval(60 * 1000L)
          // 每隔多久进行文件回滚
          .withRolloverInterval(60 * 1000L)
          // 文件大小
          .withMaxPartSize(128 * 1024 * 1024L)
          .build())
      .build()

    socketStream.map(row => row)
      .addSink(fileSink)

    streamEnv.execute("POC_Socket_Stream")
  }

}

File: hy-poc-flink/src/main/java/com/hy/poc/flink/sink/StreamingFileSinkPoc.java
package com.hy.poc.flink.sink;

import org.apache.flink.api.common.serialization.SimpleStringEncoder;
import org.apache.flink.core.fs.Path;
import org.apache.flink.runtime.state.hashmap.HashMapStateBackend;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.CheckpointConfig;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.sink.filesystem.OutputFileConfig;
import org.apache.flink.streaming.api.functions.sink.filesystem.StreamingFileSink;
import org.apache.flink.streaming.api.functions.sink.filesystem.bucketassigners.DateTimeBucketAssigner;
import org.apache.flink.streaming.api.functions.sink.filesystem.rollingpolicies.DefaultRollingPolicy;

import java.io.IOException;
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;

/**
 * StreamingFileSinkPoc
 *
 * @author Jie.Hu
 * @date 8/6/21 2:10 PM
 */
public class StreamingFileSinkPoc {

    private static String HASH_MAP_CHECKPOINT_DIR = "file:///Users/hujie/IdeaProjects/hy-learn/hy-poc-flink/out/hashmap_checkpoint";

    private static String SINK_DIR = "file:///Users/hujie/IdeaProjects/hy-learn/hy-poc-flink/out/streamfilesink";

    public static void mock() {
        Thread thread = new Thread() {
            @Override
            public void run() {
                try {
                    ServerSocket serverSocket = new ServerSocket(8989, 3);
                    while (true) {
                        Socket socket = serverSocket.accept();
                        System.out.println(socket.getInetAddress().toString().substring(1) + ":" + socket.getPort() + " Connected");
                        OutputStream output = socket.getOutputStream();
                        int count = 0;
                        while (true) {
                            output.write(String.join(",", "number", String.valueOf(count++), "jayden\n").getBytes());
                            Thread.sleep(50);
                        }
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                }

            }
        };
        thread.start();
    }

    public static void main(String[] args) throws Exception {
        // mock data
        mock();

        StreamExecutionEnvironment streamEnv = StreamExecutionEnvironment.getExecutionEnvironment();
        ///////// checkpoint /////////
        // 每 6 min 开启一次 checkpoint
        streamEnv.enableCheckpointing(6 * 60 * 1000);
        CheckpointConfig checkpointConfig = streamEnv.getCheckpointConfig();
        // 当取消任务的时候，是否保留 checkpoint 数据，默认情况下不保留
        // 请注意，当任务异常退出时是会保留 checkpoint 数据的
        checkpointConfig.enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);
        checkpointConfig.setCheckpointStorage(HASH_MAP_CHECKPOINT_DIR);
        // checkpoint 并行度，如果设置了 MinPauseBetweenCheckpoints 即 checkpoint 的最小停顿时间，checkpoint 最大并行度只能是 1
        checkpointConfig.setMaxConcurrentCheckpoints(1);
        // checkpoint 之间的最小停顿时间，
        checkpointConfig.setMinPauseBetweenCheckpoints(2 * 60 * 1000);
        // checkpoint 容错失败次数，若 checkpoint 连续失败 times 次，则重启作业
        checkpointConfig.setTolerableCheckpointFailureNumber(5);
        // checkpoint 过期时间
        checkpointConfig.setCheckpointTimeout(2 * 60 * 1000);


        ///////// state backend /////////
        streamEnv.setStateBackend(new HashMapStateBackend());

        //////// file sink ////////
        DataStreamSource<String> socketTextStream = streamEnv.socketTextStream("127.0.0.1", 8989);
        socketTextStream.name("source");

        StreamingFileSink fileSink = StreamingFileSink
                .forRowFormat(new Path(SINK_DIR), new SimpleStringEncoder<String>())
                .withBucketAssigner(new DateTimeBucketAssigner())
                .withRollingPolicy(DefaultRollingPolicy
                        .builder()
                        .withRolloverInterval(1 * 60 * 1000)
                        .withMaxPartSize(128 * 1024 * 1024L)
                        .withInactivityInterval(50 * 1000)
                        .build())
                .withOutputFileConfig(OutputFileConfig.builder()
                        .withPartPrefix("ck_row")
                        .withPartSuffix(".csv")
                        .build())
                .build();

        socketTextStream.rebalance();
        socketTextStream
                .addSink(fileSink)
                .name("sink")
                .setParallelism(4);

        streamEnv.execute();
    }
}
